### PR TITLE
Get SDL wrapper to use dynamic library loading

### PR DESCRIPTION
**Pull request by cra0zy**
To make it easier to review, first comes just SDL.

CC. @dellis1972

### CODE REVIEW

**Comment by cra0zy** [comment]
> @cra0zy The point of this is to make moving to .NET Standard easier?

Among other things, yes.

**Comment by Jjagg** [comment]
What are the other things?

**Comment by cra0zy** [comment]
> What are the other things?

More powerful library loading, look at GetNativeLibrary in SDL2.cs.

**Comment by mrhelmut** [comment]
Would this work correctly on Windows? I may be missing something, but I feel like the 32/64bit detection is gone (since there is no .dll.config on Windows .NET).

Also, just noting that bumping the requirement to .NET 4.5 will drop support for Windows XP (which is no big deal to me -- 0.20% of the Steam users -- but we need to be aware of this when proceeding; e.g. updating the docs).

Other than this, I'm all in for this. I can't wait to see the new project architecture coming together.

**Comment by cra0zy** [comment]
> Would this work correctly on Windows? I may be missing something, but I feel like the 32/64bit detection is gone (since there is no .dll.config on Windows .NET).

We are using `Environment.Is64BitProcess` to select proper libraries on Windows, and this PR has not changed that.

> Also, just noting that bumping the requirement to .NET 4.5 will drop support for Windows XP (which is no big deal to me -- 0.20% of the Steam users -- but we need to be aware of this when proceeding; e.g. updating the docs).

This PR does not change .NET Framework requirement.

File: MonoGame.Framework/Utilities/CurrentPlatform.cs [hunk 1]
```diff
@@ -99,9 +99,9 @@ public static class PlatformParameters
...
     {
         [DllImport("kernel32.dll", SetLastError = true)]
-        private static extern bool SetDllDirectory(string lpPathName);
-        
-        private static bool _dllDirectorySet = false;
+        private static extern bool SetDllDirectory(string lpPathName);
+        
+        private static bool _dllDirectorySet = false;
         
         public static void InitDllDirectory()
```
[the remark this answers is not in the document]
Comment (cra0zy) [reply MonoGame.Framework/Utilities/CurrentPlatform.cs]: Yup.

File: MonoGame.Framework/Utilities/FuncLoader.Desktop.cs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            [DllImport("libdl.so.2")]
+            public static extern IntPtr dlopen(string path, int flags);
+
+            [DllImport("libdl.so.2")]
+            public static extern IntPtr dlsym(IntPtr handle, string symbol);
+        }
+
+        private class OSX
+        {
+            [DllImport("/usr/lib/libSystem.dylib")]
```
[the remark this answers is not in the document]
Comment (cra0zy) [reply MonoGame.Framework/Utilities/FuncLoader.Desktop.cs]: I don't own a Mac, you'll have to test it and tell me... tho it doesn't really hurt.

File: MonoGame.Framework/Utilities/FuncLoader.Desktop.cs [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        private static extern int uname(IntPtr buf);
+        
+        private const int RTLD_LAZY = 0x0001;
+        private const int RTLD_GLOBAL = 0x0100;
+
+        public static bool IsWindows, IsOSX;
+
+        static FuncLoader()
+        {
+            switch (Environment.OSVersion.Platform)
```
[the remark this answers is not in the document]
Comment (cra0zy) [reply MonoGame.Framework/Utilities/FuncLoader.Desktop.cs]: Good point.

**Comment by cra0zy** [comment]
@dellis1972 All good now?

**Comment by dellis1972** [comment]
@cra0zy all good from me. @Jjagg did you want to give this one last review then merge?

**Comment by Jjagg** [comment]
Looks good! Thanks @cra0zy!

**Comment by mellinoe** [comment]
This is completely unsolicited advice (I'm just here lurking 😅 ), but you might need to apply `[UnmanagedFunctionPointer(CallingConvention.Cdecl)]` to the new delegate type declarations. I had a related bug that I fixed here: https://github.com/mellinoe/veldrid/commit/444d3295bf97346ffafdbdd0960c96fdce91d0a8. Folks were reporting problems on 32-bit Windows, where the calling convention mismatch caused problems.

**Comment by tomspilman** [comment]
> This is completely unsolicited advice 

Your advice is always appreciated, Eric!

@cra0zy @dellis1972 - Can you look into this.

**Comment by cra0zy** [comment]
Will do.

**Comment by dellis1972** [comment]
Ping me when the PR is up, and I'll get it merged asap.

**Comment by cra0zy** [comment]
https://github.com/MonoGame/MonoGame/pull/6320

**Comment by mrhelmut** [comment]
Any chance to have a PR doing the same for OpenAL-Soft? Right now this PR makes Windows builds unable to find oal_soft.dll because the dll directory is set to the default directory.

**Comment by cra0zy** [comment]
> Any chance to have a PR doing the same for OpenAL-Soft? Right now this PR makes Windows builds unable to find oal_soft.dll because the dll directory is set to the default directory.

I did not change any openal stuff, are you talking about the thing with working dir + mono + .dll.config?

And yea, I should be able to do the same for OpenAL very soon, maybe even today.

**Comment by mrhelmut** [comment]
Removing this is what made OpenAL to not be able to load anymore:

```
if (PlatformParameters.DetectWindowsArchitecture)
               NativeHelper.InitDllDirectory();
```

It was calling a system function that manually sets Windows dll discovery (since you can't use a dll mapping based on the architecture on Windows). We still need something to detect x86/x64 and tell Windows that our app needs to load dlls from the x86 or x64 folder.

What you did for SDL is cool, but we need to load OpenAL this way too if you'd like to remove ```InitDllDirectory()``` for good.

**Comment by cra0zy** [comment]
> Removing this is what made OpenAL to not be able to load anymore on Windows:

Its still there... https://github.com/MonoGame/MonoGame/blob/b914f314a22dae8530ecfd5232adc8ceec55782f/MonoGame.Framework/Audio/OpenALSoundController.cs#L113

**Comment by mrhelmut** [comment]
I missed that one! Though ```#if WINDOWS``` should be ```#if DESKTOPGL``` I believe, that's why it doesn't get triggered.

**Comment by cra0zy** [comment]
WIP branch for openal and opengl funcloader is available at: https://github.com/cra0zy/MonoGame/tree/openaldelegates

Need to fix OpenGL calling for Mac/Android/iOS and than test it out. 

Hopefully will be done tomorrow.

**Comment by mrhelmut** [comment]
So far so good! Awesome work.

**Pull request merged**